website: Add data consistency error docs for "planned value does not match config value"

austinvalle opened the pull request:
Ref: #1229 

This PR adds additional documentation about a common data consistency error we see during migration to Plugin Framework:
```bash
TIMESTAMP [WARN]  Provider "TYPE" produced an invalid plan for ADDRESS, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .ATTRIBUTE: planned value cty.StringVal("VALUE") does not match config value cty.StringVal("value")
```

I added a simplified example for demonstrating the behavior, but can adjust/add more examples if we think that'd be helpful.

bendbennett commented on website/docs/plugin/sdkv2/resources/data-consistency-errors.mdx:
```diff
@@ -110,3 +110,140 @@ This occurs if the attribute schema definition is `Optional: true` without `Comp
+	basetypes.StringValue
+}
+
+// ... custom value type implementation
```
Not sure if it's worth adding further detail about what would need to be implemented for a custom value type, or perhaps linking to the relevant docs?

austinvalle replied:
Good suggestion. I'm going to move this section over to the migration guide in the framework repo, but I'll add a `<Note>` above this code example that says this isn't a full example and links to the custom value documentation 👍🏻

austinvalle replied:
Added suggestion in new PR, see https://github.com/hashicorp/terraform-plugin-framework/pull/831/files#diff-671bcafb4ced355e43bc6d152c9bef463a5fc27fa42d2ec2eaa4b9409004d33aR245-R249

bendbennett reviewed (approved):
LGTM 🚀

bflad reviewed (approved):
I don't want to hold this excellent content up from going somewhere, but something to potentially consider here. 👍

austinvalle commented on website/docs/plugin/sdkv2/resources/data-consistency-errors.mdx:
````diff
@@ -110,3 +110,140 @@ This occurs if the attribute schema definition is `Optional: true` without `Comp
+}
+```
+
+#### Migrating to Plugin Framework
````
I opened https://github.com/hashicorp/terraform-plugin-framework/pull/831 and moved the code over in 016f347

austinvalle commented:
@bflad @bendbennett - I moved the code over to a [new PR in plugin-framework](https://github.com/hashicorp/terraform-plugin-framework/pull/831). I removed the migration section + plugin framework code from this PR and added an SDKv2 example.

I'd love a second review on the SDKv2 example I provided, as I was basing a lot of it off the Semantic Equality RFC and my sandbox testing. Want to make sure this example makes sense 😄

bflad commented on website/docs/plugin/sdkv2/resources/data-consistency-errors.mdx:
````diff
@@ -110,3 +110,100 @@ This occurs if the attribute schema definition is `Optional: true` without `Comp
+}
+```
+
+A warning log will be produced and the resulting state after applying a new resource would look like:
+```json
+{
+  //...
+  "resources": [
+    {
+      "mode": "managed",
+      "type": "examplecloud_thing",
+      "name": "this",
+      "instances": [
+        {
+          "attributes": {
+            // This is invalid! Config has this stored as "value"
+            "word": "VALUE"
+          },
+        }
+      ]
+    }
+  ],
+}
+```
````
I'm not sure folks will be used to the seeing the JSON state, so maybe best to just say what will happen:

```suggestion
A warning log will be produced and the resulting state after applying a new resource it will store the attribute state as `VALUE` instead of `value`.
```

austinvalle replied:
Good point -> ⚡

bflad reviewed (approved):
This looks good to me 🚀
